File: c/src/accessdb.c
/*
 * accessdb.c — OCXcip_AccessTagDataDb (batched R/W via cached db_handle).
 *
 * SPDX-License-Identifier: MIT
 *
 * Peer of access.c (OCXcip_AccessTagData) that trades the 251-byte
 * per-call path string for a 4-byte db_handle lookup on the engine
 * side.  The handle is the value CreateTagDbHandle returned (cached
 * on bp_tagdb_t at open time).  See docs/access-tag-data-db.md for
 * the RE notes and the descriptor-layout diff vs AccessTagData.
 *
 * SDK policy on the new fields:
 *   has_extra = 0, opt_value  = 0      (matches OEM wrapper's
 *                                       param_5 == NULL branch)
 *   has_data  = 0, mask_seed  = 0      (matches OEM wrapper's
 *                                       request->data == NULL branch)
 *
 * The engine reads tag data from the slot's data area when has_data == 0,
 * which is the same path AccessTagData already drives successfully.
 */
#include <stdint.h>
#include <string.h>

#include "../include/bpclient.h"
#include "proto.h"

typedef struct {
    bp_tagdb_t       *db;
    bp_tag_request_t *requests;
    size_t            count;
    uint32_t          data_area_start;  /* slot offset where data area begins */
} accessdb_ctx_t;

static void accessdb_fill(uint8_t *slot, void *user) {
    accessdb_ctx_t *ctx = user;

    /* db_handle (u32) at +0x78; has_extra/opt_value/count immediately
     * after.  Zero the run from +0x78 through the end of the header
     * area so any reserved bytes start at 0. */
    memset(slot + BP_HDR_PAYLOAD_START, 0,
           BP_TAGDATA_DB_REQ0_START - BP_HDR_PAYLOAD_START);

    bp_st_u32(slot + BP_TAGDATA_DB_HANDLE_OFF, ctx->db->handle);
    /* has_extra/opt_value already 0 from the memset above. */
    bp_st_u16(slot + BP_TAGDATA_DB_COUNT_OFF, (uint16_t)ctx->count);

    /* Per-request descriptors + inline data. */
    uint32_t data_off = ctx->data_area_start;
    for (size_t i = 0; i < ctx->count; i++) {
        bp_tag_request_t *r = &ctx->requests[i];
        uint32_t req_start =
            BP_TAGDATA_DB_REQ0_START + (uint32_t)i * BP_TAGDATA_DB_REQ_STRIDE;

        /* Zero the whole descriptor, then populate fields.  Names up
         * to 254 bytes are copied; the NUL at +0xFF stays from the
         * memset (the OEM wrapper writes an explicit 0 there — same
         * effect since we memset first). */
        memset(slot + req_start, 0, BP_TAGDATA_DB_REQ_STRIDE);

        size_t tlen = strlen(r->tag_name);
        if (tlen > 254) tlen = 254;
        memcpy(slot + req_start + BP_REQDB_TAGNAME_OFF, r->tag_name, tlen);

        bp_st_u16(slot + req_start + BP_REQDB_ACTION_OFF,         r->action);
        bp_st_u32(slot + req_start + BP_REQDB_DATATYPE_OFF,       r->data_type);
        bp_st_u32(slot + req_start + BP_REQDB_ELEM_BYTE_SIZE_OFF, r->elem_byte_size);
        bp_st_u32(slot + req_start + BP_REQDB_ELEM_COUNT_OFF,     r->elem_count);
        /* has_data and mask_seed both stay at 0 from the memset. */

        /* For writes, copy caller's bytes into the slot's data area. */
        uint32_t nbytes = (uint32_t)r->elem_byte_size * (uint32_t)r->elem_count;
        if (r->action == BP_ACTION_WRITE && r->data && nbytes > 0) {
            memcpy(slot + data_off, r->data, nbytes);
        }
        data_off += nbytes;
    }
}

static void accessdb_read(uint8_t *slot, void *user) {
    accessdb_ctx_t *ctx = user;
    uint32_t data_off = ctx->data_area_start;
    for (size_t i = 0; i < ctx->count; i++) {
        bp_tag_request_t *r = &ctx->requests[i];
        uint32_t req_start =
            BP_TAGDATA_DB_REQ0_START + (uint32_t)i * BP_TAGDATA_DB_REQ_STRIDE;
        r->result = bp_ld_u32(slot + req_start + BP_REQDB_RESULT_OFF);
        uint32_t nbytes = (uint32_t)r->elem_byte_size * (uint32_t)r->elem_count;
        if (r->action == BP_ACTION_READ && r->data && nbytes > 0) {
            memcpy(r->data, slot + data_off, nbytes);
        }
        data_off += nbytes;
    }
}

int bp_tagdb_access_db(bp_tagdb_t *db, bp_tag_request_t *requests, size_t count) {
    if (!db || !requests || count == 0) return BP_ERR_NULL_ARG;
    if (count > 16)                      return BP_ERR_PARAM_RANGE;

    /* data_area_start = 0x1B0 + (count - 1) * 0x128, matching the
     * OEM wrapper's payload_size = 0x1B0 + (count-1)*0x128 +
     * total_data_bytes formula. */
    uint32_t data_area_start =
        BP_TAGDATA_DB_DATA_AREA0
        + ((uint32_t)count - 1u) * BP_TAGDATA_DB_REQ_STRIDE;
    uint32_t total_data_bytes = 0;
    for (size_t i = 0; i < count; i++) {
        total_data_bytes += (uint32_t)requests[i].elem_byte_size
                          * (uint32_t)requests[i].elem_count;
    }
    uint32_t payload_size = data_area_start + total_data_bytes;
    if (payload_size > BP_SLOT_STRIDE - 0x80) return BP_ERR_SLOT_TOO_LARGE;

    accessdb_ctx_t ctx = {
        .db = db, .requests = requests, .count = count,
        .data_area_start = data_area_start,
    };
    bp_call_spec_t spec = {
        .fn_name      = "OCXcip_AccessTagDataDb",
        .payload_size = payload_size,
        .fill_payload = accessdb_fill,
        .read_reply   = accessdb_read,
        .timeout_ms   = 10000,
        .user         = &ctx,
    };
    return bp_client_call(db->client, &spec);
}

File: go/ocxbp/cip/accessdb.go
// SPDX-License-Identifier: MIT

package cip

import "encoding/binary"

// AccessTagDataDb layout constants (v0.10.4+). Peer of AccessTagData
// using the cached db_handle from CreateTagDbHandle instead of the
// path string. Wire layout — including the +8 descriptor stride
// (0x128 vs 0x120), the widened u32 data_type/elem_byte_size/
// elem_count fields, and the new has_data / mask_seed pair — is
// described in docs/access-tag-data-db.md. Mirrors c/src/proto.h.
const (
	TagDataDbHandleOff   = 0x78  // uint32 db_handle
	TagDataDbHasExtraOff = 0x7C  // uint8  has_extra (we ship 0)
	TagDataDbOptValueOff = 0x7E  // uint16 opt_value (we ship 0)
	TagDataDbCountOff    = 0x80  // uint16 request count
	TagDataDbReq0Start   = 0x88  // first descriptor
	TagDataDbReqStride   = 0x128 // bytes per descriptor
	TagDataDbDataArea0   = 0x1B0 // data area start when count == 1

	// Within an AccessTagDataDb request descriptor (relative to
	// descriptor start). Field offsets and widths DIFFER from
	// AccessTagData — see the table in docs/access-tag-data-db.md.
	ReqDbTagNameOff      = 0x000 // char[255] + NUL
	ReqDbActionOff       = 0x100 // uint16 1=read 2=write
	ReqDbDataTypeOff     = 0x104 // uint32 (widened from u16)
	ReqDbElemByteSizeOff = 0x108 // uint32 (widened from u16)
	ReqDbElemCountOff    = 0x10C // uint32 (widened from u16)
	ReqDbHasDataOff      = 0x110 // uint8  (we ship 0)
	ReqDbMaskSeedOff     = 0x118 // uint64 (we ship 0)
	ReqDbResultOff       = 0x120 // uint32 server-written
)

// AccessTagDataDbPayloadSize returns the per-call payload_size for
// OCXcip_AccessTagDataDb. See docs/protocol.md "OCXcip_AccessTagDataDb".
func AccessTagDataDbPayloadSize(reqs []TagRequest) uint32 {
	dataAreaStart := uint32(TagDataDbDataArea0)
	if n := len(reqs); n > 1 {
		dataAreaStart += uint32(n-1) * TagDataDbReqStride
	}
	var totalData uint32
	for i := range reqs {
		totalData += uint32(reqs[i].ElemByteSize) * uint32(reqs[i].ElemCount)
	}
	return dataAreaStart + totalData
}

// EncodeAccessTagDataDb writes the db_handle, count, per-request
// descriptors, and outbound write data into the slot. Mirrors
// EncodeAccessTagData but for the AccessTagDataDb wire format —
// the descriptor stride, field offsets, and field widths all differ.
//
// has_extra, opt_value, has_data, and mask_seed are all left at 0
// (matching the OEM wrapper's NULL-pointer branches; engine reads
// data from the slot data area, which we populate inline below).
func EncodeAccessTagDataDb(slot []byte, dbHandle uint32, reqs []TagRequest) {
	binary.LittleEndian.PutUint32(slot[TagDataDbHandleOff:], dbHandle)
	binary.LittleEndian.PutUint16(slot[TagDataDbCountOff:], uint16(len(reqs)))
	// has_extra (byte) and opt_value (u16) stay 0 from the
	// dispatcher's bulk-clear of the slot.

	dataOff := uint32(TagDataDbDataArea0)
	if n := len(reqs); n > 1 {
		dataOff += uint32(n-1) * TagDataDbReqStride
	}

	for i := range reqs {
		r := &reqs[i]
		reqStart := TagDataDbReq0Start + uint32(i)*TagDataDbReqStride

		// tag_name (max 254 bytes + NUL terminator at +0xFF; the
		// terminator stays 0 from the dispatcher's bulk-clear).
		tn := len(r.TagName)
		if tn > 254 {
			tn = 254
		}
		copy(slot[reqStart+ReqDbTagNameOff:reqStart+ReqDbTagNameOff+uint32(tn)], r.TagName)

		binary.LittleEndian.PutUint16(slot[reqStart+ReqDbActionOff:], r.Action)
		binary.LittleEndian.PutUint32(slot[reqStart+ReqDbDataTypeOff:], uint32(r.DataType))
		binary.LittleEndian.PutUint32(slot[reqStart+ReqDbElemByteSizeOff:], uint32(r.ElemByteSize))
		binary.LittleEndian.PutUint32(slot[reqStart+ReqDbElemCountOff:], uint32(r.ElemCount))
		// has_data and mask_seed stay 0 from the dispatcher's clear.

		nbytes := uint32(r.ElemByteSize) * uint32(r.ElemCount)
		if r.Action == ActionWrite && len(r.Data) > 0 && nbytes > 0 {
			cnt := nbytes
			if uint32(len(r.Data)) < cnt {
				cnt = uint32(len(r.Data))
			}
			copy(slot[dataOff:dataOff+cnt], r.Data[:cnt])
		}
		dataOff += nbytes
	}
}

// DecodeAccessTagDataDb reads per-request result codes (at +0x120,
// not +0x118 as for AccessTagData) and copies inbound read data
// from the slot data area back into each request's Data buffer.
func DecodeAccessTagDataDb(slot []byte, reqs []TagRequest) {
	dataOff := uint32(TagDataDbDataArea0)
	if n := len(reqs); n > 1 {
		dataOff += uint32(n-1) * TagDataDbReqStride
	}
	for i := range reqs {
		r := &reqs[i]
		reqStart := TagDataDbReq0Start + uint32(i)*TagDataDbReqStride
		r.Result = binary.LittleEndian.Uint32(slot[reqStart+ReqDbResultOff:])
		nbytes := uint32(r.ElemByteSize) * uint32(r.ElemCount)
		if r.Action == ActionRead && len(r.Data) > 0 && nbytes > 0 {
			cnt := nbytes
			if uint32(len(r.Data)) < cnt {
				cnt = uint32(len(r.Data))
			}
			copy(r.Data[:cnt], slot[dataOff:dataOff+cnt])
		}
		dataOff += nbytes
	}
}
